MSC4198: Usage of OIDC login_hint

tonkku107 opened the pull request:
[Rendered](https://github.com/tonkku107/matrix-spec-proposals/blob/oidc-login-hint/proposals/4198-oidc-login-hint.md)

**Implementations:**
- Authorization code flow
  - Homeserver / Authentication service:
    - https://github.com/element-hq/matrix-authentication-service/pull/3343
  - Client:
    - Element X
      - https://github.com/element-hq/element-x-ios/pull/4108
      - https://github.com/element-hq/element-x-android/pull/4752
 - Device authorization flow
    - Homeserver / Authentication service:
      - TBD
    - Client:
      - TBD
 - Account management requests
   - Homeserver / Authentication service:
     - https://github.com/element-hq/matrix-authentication-service/pull/5516
   - Client:
     - TBD
 - Metadata
   - Homeserver / Authentication service:
     - TBD

tonkku107 committed:
First draft

tonkku107 committed:
Rename

Johennes commented on proposals/4198-oidc-login-hint.md:
```diff
@@ -0,0 +1,131 @@
+To improve the UX of this flow, the MXID can be sent to the homeserver with the authorization request in the `login_hint`.
+In order to comply with the OpenID Connect specification, the requested scope must also include the `openid` scope.
+
+The value of `login_hint` should be set to the user's MXID, prefixed with `mxid:` (e.g. `mxid:@example-user:example.com`).
```
I'm trying to understand if the prefix is meant to be mandatory or not. Is "should" here the RFC2119 SHOULD and does it relate to inserting the MXID, the `mxid:` prefix or both?

tonkku107 replied:
The prefix is meant to be mandatory, but the inclusion of a value at all is optional, just like in openid if that makes sense. I didn't strictly use RFC2119 keywords (which is why they're not in caps) but I ended up writing in that style, probably thanks to spending too much time reading RFCs

Johennes replied:
Ok, gotcha and that sounds sensible. I probably just read it a little to hairsplitting.

tonkku107 replied:
The latest commit should clarify this

Johennes replied:
Thank you. 👍

Johennes commented on proposals/4198-oidc-login-hint.md:
```diff
@@ -0,0 +1,131 @@
+In order to comply with the OpenID Connect specification, the requested scope must also include the `openid` scope.
+
+The value of `login_hint` should be set to the user's MXID, prefixed with `mxid:` (e.g. `mxid:@example-user:example.com`).
+Prefixing the value allows for more hint value types to be added in the future and for easy detection between them.
```
Just curious, do you already foresee any types of identifiers that would require differentiation via a prefix?

clokep replied:
Maybe you'd want email in some situations?

Johennes replied:
Ok, makes sense. Might help to spell this out in the proposal.

Johennes replied:
It's possibly just me and I might be overthinking this but I couldn't find any common standards around login hint prefixes so would have welcomed some verbosity of why we need them or what grammar they might require.

tonkku107 replied:
Yeah, I agree the grammar should be explained in more detail, even if it's just `<anything>:<something>`. I do feel like that one sentence already explains why you would want prefixes quite well but maybe I'm overlooking something.

Johennes replied:
I just failed to think of a colliding example because I didn't consider ActivityPub. But if I'm the only one, then no need making that explicit. 😅

tonkku107 replied:
I've done some restructuring to add more detail for the format

tonkku107 committed:
Clarify prefix requirement

tonkku107 committed:
Specify the format in more detail

tonkku107 committed:
Remove OpenID requirement

pixlwave commented on proposals/4198-oidc-login-hint.md:
```diff
@@ -0,0 +1,156 @@
+
+#### Passing a login_hint to upstream providers
+
+To take full advantage of the benefits of `login_hint`, a suitable value could be forwarded to an upstream provider.
+The vague nature of the values for `login_hint` could make this difficult to implement, however.
```
Reading this, I'm questioning whether the `mxid:` prefix is helpful or a hinderance.

If for example an upstream provider is expecting a plain username (or their own internal identifier format), what would be the trade-off with supplying the raw hint without a prefix and letting whichever provider (upstream or not) parse that (or throw) however it sees fit? When you consider that the prefix is a hint for a hint, that starts to set alarm bells ringing for me.

The specific use case I'm thinking of here is that some clients already support onboarding by opening a URL such as `client.com/login?homeserver=myserver.net`. If the client wanted to add an additional `login_hint=some_username` query parameter then the `mxid:` adds the following friction points:
- As we add more supported ID types, each server admin would need to check the spec to see if the type they're using is supported before provisioning onboarding links.
- If the type of ID they would like to use doesn't exist (or is hard to derive for the existing types) then they'll need to open a new proposal to add it (which would block onboarding their users).
- Ultimately it seems like someone would then create an `other:` prefix as a catch all, at which point we may as well drop the prefix and just let the provider interpret the hint however it sees fit.

tonkku107 replied:
It is not expected that the upstream provider supports the `mxid` hint. It'd have to be up to the person configuring the upstream provider on their homeserver to figure out what type of value the upstream provider expects.

pixlwave replied:
Discussed this with @sandhose, the conclusion was that this MSC doesn't explicitly prevent a value without the `mxid:*` hint so there isn't a concern here :)

tonkku107 committed:
Reword upstream provider section

noelportillo reviewed (commented):
First thing that came out

hughns reviewed (changes requested):
I suggest:

- Updating to reflect that OAuth 2.0 API now exists in the spec
- Making it explicit that this is an optional parameter for the Authorization code grant
- Propose that that the `login_hint` can also be added to the MSC4191 account management URL query parameters (perhaps prefixed as `org.matrix.msc4198.login_hint`)

tonkku107 committed:
Adapt to stable OAuth 2.0 API

tonkku107 committed:
Make requirements and optionality clearer

By using RFC2119 keywords and reiterating that the parameter is optional in the usage section

tonkku107 committed:
Expand to MSC4191

hughns commented on proposals/4198-oidc-login-hint.md:
```diff
@@ -0,0 +1,180 @@
+TBD
+
+
+## Alternatives
```
Another obvious alternative is to use the bare MXID: `login_hint=@example-user:example.com`.

tonkku107 replied:
That was technically accounted for with the mention of the domain being redundant to the server but now that I look at it again the section could use some rewording.

tonkku107 replied:
Reworded in aa67dc0

hughns commented on proposals/4198-oidc-login-hint.md:
```diff
@@ -0,0 +1,180 @@
+
+### Format
+
+In order to futureproof for new new types of hints in the future and to be able to identify and parse them easily,
```
How can a client determine which types of hint a server can support?

Should we add something like a `login_hint_types_supported[]` to the [discovery](https://spec.matrix.org/v1.15/client-server-api/#server-metadata-discovery) response?

e.g. Imagining a future MSC adds `email:example-user@example.com` and `localpart:example-user` hint types:

```
{
...
	"login_hint_types_supported": ["mxid", "email", "localpart"],
...
}
```

tonkku107 replied:
That's a good point. We probably need to call it `org.matrix.login_hint_types_supported` as it wouldn't make sense to officially register this like we did with `account_management_uri` and `account_management_actions_supported`

tonkku107 replied:
Added in a6c1187

hughns commented on proposals/4198-oidc-login-hint.md:
```diff
@@ -0,0 +1,180 @@
+
+## Dependencies
+
+This MSC builds on [MSC4191] (which at the time of writing has not yet been accepted into the spec).
```
MSC4191 has now been accepted.

tonkku107 replied:
Done in 5f215af

hughns commented on proposals/4198-oidc-login-hint.md:
```diff
@@ -0,0 +1,180 @@
+uses the wrong credentials.
+The client MAY inform the user about ending up on a different account than intended and present an option to try again.
+
+#### Example authorization request
```
Could we also make this apply to the [device authorisation flow](https://spec.matrix.org/v1.19/client-server-api/#device-authorisation-flow)?

i.e. in the Device Authorisation Request POST request allow the `login_hint` param to be passed.

tonkku107 replied:
Most other parameters also get forgotten about for the device authorization flow in the oauth world. It might be worth looking into something in the future that defines common parameters available in every flow in the spec.

RFC8628 doesn't seem to make any mention of how to go about extending the parameters but it's probably fine to just add `login_hint` as is since that's what OIDC already did for the authorization code flow.

tonkku107 replied:
Added in 491d18b

hughns commented on proposals/4198-oidc-login-hint.md:
```diff
@@ -0,0 +1,180 @@
+
+To take full advantage of the benefits of `login_hint`, a suitable value that the upstream provider expects could be
+supplied when logging in with one.
+Since this proposal only covers the Matrix protocol, upstream providers likely won't follow the format laid out in this
```
Maybe this suggests an alternative to consider: that we use a format that is more common with upstream providers?

If upstreams commonly accept a plain email or username then does that suggest we might want to as well? e.g. `login_hint=example-user@example.com` and `login_hint=example-user`

tonkku107 replied:
As stated in the introductory section, there is no defined format and every upstream is going to have their own idea of how to use it. There is no guarantee either that the email or username used on the upstream account matches the localpart or email set on someone's matrix account. It's best to leave what's sent to the upstream as a configurable option on the homeserver.

tonkku107 replied:
Mentioned that in aa67dc0

tonkku107 committed:
Updates with recent versions of the spec

tonkku107 committed:
New alternative

tonkku107 committed:
Add metadata field

tonkku107 committed:
Move MXID bits out of the auth code flow section

tonkku107 committed:
Add support for device authorization flow
